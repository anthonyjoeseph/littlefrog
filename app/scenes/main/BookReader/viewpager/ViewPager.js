// app/BookReader/ViewPager.js

import React, { Component } from 'react';
import {
  StyleSheet,
  View,
  Text,
  Image,
  ScrollView
} from 'react-native';

function isNearInt(value, err) {
  var nearbyInt = Math.round(value);
  return Math.abs(value - nearbyInt) < 0.05;
}

class ViewPager extends Component {

  constructor(props){
    super(props);
    this.currentPage = 0;

    var pagesShouldRender = [];
    for(var i = 0; i < props.numPages; i++){
      if(i <= this.props.renderRange){
        pagesShouldRender[i] = true;
      }else{
        pagesShouldRender[i] = false;
      }
    }
    this.state = {
      pagesShouldRender: pagesShouldRender
    };

    this.onScroll = this.onScroll.bind(this);
  }

  render(){
    return (
      <ScrollView
        horizontal={true}
        pagingEnabled={true}
        scrollEventThrottle={1}
        onScroll={this.onScroll}>
          {
            React.Children.map(
              this.props.children,
              function(child, index){
                return React.cloneElement(child, {
                  shouldRender: this.state.pagesShouldRender[index]
                })
              }.bind(this)
            )
          }
      </ScrollView>
    );
  }

  updateRenderedPages(){
    var currentPage = this.currentPage;
    var pagesShouldRender = [];
    for(var i = 0; i < this.props.numPages; i++){
      if(i < currentPage - this.props.renderRange){
        pagesShouldRender[i] = false;
      }else if(i <= currentPage + this.props.renderRange){
        pagesShouldRender[i] = true;
      }else{
        pagesShouldRender[i] = false;
      }
    }
    this.setState({pagesShouldRender: pagesShouldRender});
  }

  onScroll(event){
    var fullLength = event.nativeEvent.contentSize.width;
    var pageWidth = Math.floor(fullLength / this.props.numPages);
    var x = event.nativeEvent.contentOffset.x;
    var pageFraction = x / pageWidth;
    if(isNearInt(pageFraction)){
      var newPage = Math.round(pageFraction);
      if(this.state.currentPage !== newPage){
        this.props.onPageScroll(newPage);
        this.currentPage = newPage;
        this.updateRenderedPages();
      }
    }else{
      if(pageFraction > this.currentPage){
        if(this.currentPage + 1 < this.props.numPages){
          this.props.onPageScroll(this.currentPage + 1);
        }
      }else if(pageFraction > 0){
        this.props.onPageScroll(this.currentPage - 1);
      }
    }
  }
}
export default ViewPager;
