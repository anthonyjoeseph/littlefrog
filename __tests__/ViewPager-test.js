// __tests__/DismissableFullscreenPopup-test.js
import {View, Text} from 'react-native';
import React from 'react';
import ViewPager from '../app/scenes/main/BookReader/viewpager/ViewPager';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(
    <ViewPager
      numPages={9}
      renderRange={3}
      onPageScroll={ () => {} }>
      <View><Text>Testing 1</Text></View>
      <View><Text>Testing 2</Text></View>
      <View><Text>Testing 3</Text></View>
      <View><Text>Testing 4</Text></View>
      <View><Text>Testing 5</Text></View>
      <View><Text>Testing 6</Text></View>
      <View><Text>Testing 7</Text></View>
      <View><Text>Testing 8</Text></View>
      <View><Text>Testing 9</Text></View>
    </ViewPager>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
