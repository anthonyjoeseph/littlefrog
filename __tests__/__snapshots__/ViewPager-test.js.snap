exports[`test renders correctly 1`] = `
<ScrollView
  horizontal={true}
  onScroll={[Function]}
  pagingEnabled={true}
  scrollEventThrottle={1}>
  <View
    shouldRender={true}>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Testing 1
    </Text>
  </View>
  <View
    shouldRender={true}>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Testing 2
    </Text>
  </View>
  <View
    shouldRender={true}>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Testing 3
    </Text>
  </View>
  <View
    shouldRender={true}>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Testing 4
    </Text>
  </View>
  <View
    shouldRender={false}>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Testing 5
    </Text>
  </View>
  <View
    shouldRender={false}>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Testing 6
    </Text>
  </View>
  <View
    shouldRender={false}>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Testing 7
    </Text>
  </View>
  <View
    shouldRender={false}>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Testing 8
    </Text>
  </View>
  <View
    shouldRender={false}>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Testing 9
    </Text>
  </View>
</ScrollView>
`;
